text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+
+```rust
+struct Type1;
+struct Type2;
+
+trait Trait1 {
+    fn trait1() {
+        println!("1");
+    }
+}
+impl<T> Trait1 for T {} // <--
+
+trait Trait2 {
+    fn trait2() {
+        println!("2");
+    }
+}
+impl Trait2 for Type2 {} // <--
+
+trait Say {
+    fn say();
+}
+impl<T: Trait1> Say for T
+where
+    T: Trait1, // <--
+{
+    fn say() {
+        T::trait1();
+    }
+}
+
+{
+    use impl<T> Say for T
+    where
+        T: Trait2 // <--
+    {
+        fn say() {
+            T::trait2();
+        }
+    }
+
+    Type1::say(); // 1
+    Type2::say(); // 2
+}
+```
+
+## No priority over type-associated methods
+
+Scoped `impl Trait for Type` has *the same* method resolution priority as an equivalent global implementation would have if it was visible for method-binding in that scope. This means that directly type-associated functions still bind with higher priority than those available through scoped implementations.
+
+## Coercion to trait objects
+
+Due to the coercion into a trait object in the following code, the scoped implementation becomes attached to the value through the pointer meta data. This means it can then be called from other scopes:
+
+```rust
+use std::fmt::{self, Display, Formatter};
+
+fn function() -> &'static dyn Display {
+    use impl Display for () {
+        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+            write!(f, "scoped")
+        }
+    }
+
+    &()
+}
+
+println!("{}", function()); // "scoped"
+```
+
+This behaves exactly as a global implementation would.
+
+Note that the [`DynMetadata<dyn Display>`]s of the reference returned above and one that uses the global implementation would compare as distinct even if both are "`&()`".
+
+[`DynMetadata<dyn Display>`]: https://doc.rust-lang.org/stable/core/ptr/struct.DynMetadata.html
+
+## Interaction with return-position `impl Trait`
+
+Consider the following functions:
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+    use impl Trait for () {}
+
+    () // Binds on trailing `()`-expression.
+}
+
+fn function2() -> impl Trait {
+    use impl Trait for () {}
+
+    {} // Binds on trailing `{}`-block used as expression.
+}
+```
+
+In this case, the returned opaque types use the respective inner scoped implementation, as it binds on the `()` expression.
+
+The following functions do not compile, as the implicitly returned `()` is not stated *inside* the scope where the implementation is available:
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+                 ^^^^^^^^^^
+    use impl Trait for () {}
+    ---------------------
+
+    // Cannot bind on implicit `()` returned by function body without trailing *Expression*.
+}
+
+fn function2() -> impl Trait {
+                  ^^^^^^^^^^
+    use impl Trait for () {}
+    ---------------------
+
+    return; // Cannot bind on `return` without expression.
+    -------
+}
+```
```

Comment:
In theory it could be made to, but I (subectively) think that would be hard to follow, since for example other items nested in function bodies aren't available to their signature (if I'm not completely mistaken).

The function body block isn't visually located inside itself, so to me it appear one or a half level of scoping higher than a scoped `impl Trait for Type` placed inside it.

Reply:
This seems like too subtle of a distinction to make. I don't think most rust programmers conceptualize the implicit return value to be any different than an explicit return value. While, yes, there are already subtle differences with regards to implicit drop and lifetimes, I don't think this RFC should add to those pain points by distinguishing between implicit return, return with expression, and return without expression.

Reply:
I don't have a strong preference in this regard. `return` without expression at least could definitely be defined to pick up the environment on the `return` token instead without it feeling odd in the long run, so I'll change that here when I can.

If I had totally free reign, I would probably put off the matter with the fully implicit return until the very last minute and then run a survey, but I don't know if that's *a thing* 😅
It's really hard for me to tell whether reaching the end of the function body should observe the implementation environment just outside or just inside the function, though. From a purely practical/convenience point of view, I think I would say 'inside' too, and I now suspect it may also be closer to developer intent in mixed scenarios, so I'll change it to that.

(Sorry, today I'm having some trouble with the wider reason it took me so long to submit this in the first place, so I'll be a bit slow to implement changes and respond to the larger points you brought up.)